oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/oidc/web/OidcClientRegistrationEndpointFilter.java

```diff
@@ -0,0 +1,225 @@
+import org.springframework.security.web.util.matcher.AntPathRequestMatcher;
+import org.springframework.security.web.util.matcher.RequestMatcher;
+import org.springframework.util.Assert;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.List;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+/**
+ * @author Ovidiu Popa
+ * @since 0.1.1
+ */
+public class OidcClientRegistrationEndpointFilter extends OncePerRequestFilter {
+
+	private static final String SCOPE_CLAIM_DELIMITER = " ";
+
+	private final OidcClientRegistrationHttpMessageConverter clientRegistrationHttpMessageConverter =
+			new OidcClientRegistrationHttpMessageConverter();
+	private final RegisteredClientRepository registeredClientRepository;
+	private final OidcClientRegistrationToRegisteredClientConverter oidcClientToRegisteredClientConverter =
+			new OidcClientRegistrationToRegisteredClientConverter();
+	private final RegisteredClientToOidcClientRegistrationConverter registeredClientToOidcClientConverter =
+			new RegisteredClientToOidcClientRegistrationConverter();
+	private final HttpMessageConverter<OAuth2Error> errorHttpResponseConverter =
+			new OAuth2ErrorHttpMessageConverter();
+	private final RequestMatcher requestMatcher;
+	private final OAuth2AuthorizationService authorizationService;
+
+	public OidcClientRegistrationEndpointFilter(RegisteredClientRepository registeredClientRepository,
+			OAuth2AuthorizationService authorizationService, String oidcClientRegistrationUri) {
+		Assert.notNull(registeredClientRepository, "registeredClientRepository cannot be null");
+		Assert.notNull(authorizationService, "authorizationService cannot be null");
+		Assert.hasText(oidcClientRegistrationUri, "oidcClientRegistrationUri cannot be empty");
+		this.registeredClientRepository = registeredClientRepository;
+		this.authorizationService = authorizationService;
+		this.requestMatcher = new AntPathRequestMatcher(
+				oidcClientRegistrationUri,
+				HttpMethod.POST.name()
+		);
+	}
+
+	@Override
+	protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
+			throws ServletException, IOException {
+
+		if (!this.requestMatcher.matches(request)) {
+			filterChain.doFilter(request, response);
+			return;
+		}
+
+		Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+
+		if (!isPrincipalAuthenticated(authentication)) {
+			// Pass through the chain with the expectation that the authentication process
+			// will commence via AuthenticationEntryPoint
+			filterChain.doFilter(request, response);
+			return;
+		}
+
+		try {
+			validateAuthentication(authentication);
```

Comment:
Let's move all authentication logic into the new `OidcClientRegistrationAuthenticationProvider`. This `AuthenticationProvider` should ultimately check the access rights (`scope`) to see if the registration request is allowed.

Take a look at `OAuth2TokenRevocationEndpointFilter` and `OAuth2TokenRevocationAuthenticationProvider` as the implementation will be somewhat similar.